### PR TITLE
[Consensus] Garbage Collection - 2

### PR DESCRIPTION
**Pull request by akichidis**
## Description 

This is the second part of Garbage Collection which implements the following ticked next steps as outlined on the previous PR

- [x] BlockManager to respect the gc_round when accepting blocks and trigger clean ups for new gc rounds
- [x] Skip blocks that are received which are <= gc_round
- [x] Not propose ancestors that are <= gc_round
- [x] Subscriber to ask for blocks from `gc_round` when `last_fetched_round < gc_round` for a peer to prevent us from fetching unnecessary blocks

Next steps:

- [ ] Re-propose GC'ed transactions (probably not all of them)
- [ ] Implement new timestamp approach so ancestor verification is not needed
- [ ] Harden testing for GC & edge cases

## Test plan 

CI/PT

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: consensus/core/src/block_manager.rs [hunk 1]
```diff
@@ -114,69 +114,15 @@ impl BlockManager {
...
-            // ancestors do not get suspended.
-            let blocks_to_accept: Vec<_> = blocks_to_accept.into_values().collect();
-            self.dag_state
-                .write()
-                .accept_blocks(blocks_to_accept.clone());
+            let unsuspended_blocks = self.try_unsuspend_children_blocks(block.reference());
 
+            // Verify block timestamps
+            let blocks_to_accept = self
+                .verify_block_timestamps_and_accept(iter::once(block).chain(unsuspended_blocks));
```
Comment (akichidis) [inline consensus/core/src/block_manager.rs:125]: Practically most of the previous code has been extracted to the `verify_block_timestamps_and_accept` method. This won't be needed once we refactor to the new timestamp approach

File: consensus/core/src/block_manager.rs [hunk 2]
```diff
@@ -114,69 +114,15 @@ impl BlockManager {
                     missing_blocks.extend(ancestors_to_fetch);
                     continue;
                 }
-                TryAcceptResult::Processed => continue,
+                TryAcceptResult::Processed | TryAcceptResult::Skipped => continue,
```
Comment (akichidis) [inline consensus/core/src/block_manager.rs:117]: I've made the decision to let received blocks with `block.round <= gc_round` to be allowed received up to here and not straight rejected on the `AuthorityService` in order to make sure that:
* [highest_received_rounds](https://github.com/MystenLabs/sui/blob/c092832b3cc5d0c0eb08ee23ccb50a6af3e1c15f/consensus/core/src/core_thread.rs#L169) keeps getting updated. 
* metrics in the block manager for highest received verified block etc getting updated

File: consensus/core/src/block_verifier.rs [hunk 3]
```diff
@@ -185,20 +188,44 @@ impl BlockVerifier for SignedBlockVerifier {
...
-        ancestors: &[VerifiedBlock],
+        ancestors: &[Option<VerifiedBlock>],
+        gc_enabled: bool,
+        gc_round: Round,
     ) -> ConsensusResult<()> {
-        assert_eq!(block.ancestors().len(), ancestors.len());
-        // This checks the invariant that block timestamp >= max ancestor timestamp.
-        let mut max_timestamp_ms = BlockTimestampMs::MIN;
-        for (ancestor_ref, ancestor_block) in block.ancestors().iter().zip(ancestors.iter()) {
-            assert_eq!(ancestor_ref, &ancestor_block.reference());
```
Comment (mwtian) [inline consensus/core/src/block_verifier.rs:194]: I think you just need to add this line before the assert, something like:
```
let Some(ancestor_block) = ancestor_block else {
   assert!(gc_enabled && ancestor_ref.round <= gc_round);
   continue;
};
```

Comment (akichidis) [reply consensus/core/src/block_verifier.rs:194]: Yep can do, but I feel safer at this point just separating those two paths completely. The timestamp logic will get refactored anyways.

File: consensus/core/src/core.rs [hunk 4]
```diff
@@ -731,6 +741,18 @@ impl Core {
                 ancestors
                     .into_iter()
                     .filter(|block| block.author() != self.context.own_index)
+                    .filter(|block| {
+                        if gc_enabled {
```
[the remark this answers is not in the document]
Comment (akichidis) [reply consensus/core/src/core.rs:745]: > For simplicity, not use any of these filtering should work?

Well mostly for network hygiene. No reason to link to blocks that will not be committed. Not the end of the world if we don't have this for sure.

> Or check that gc depth > 1 somewhere else, and not have the block.round() == GENESIS_ROUND condition?

Hhm not sure how `gc_depth` is involved, but It's the `block.round() > gc_round` that is the "issue". `gc_round` by default is `0`, as `GENESIS_ROUND`. Alternatively the `gc_round` could be an `Optional` which for as long as it remains `None` it's practically the same as GC being disabled. Once it has `Some` value then we start taking it into account . But not sure we want to do that.

File: consensus/core/src/subscriber.rs [hunk 5]
```diff
@@ -59,11 +59,20 @@ impl<C: NetworkClient, S: NetworkService> Subscriber<C, S> {
...
+        let mut last_received = dag_state.get_last_block_for_authority(peer).round();
+        let gc_round = dag_state.gc_round();
+        let gc_enabled = dag_state.gc_enabled();
+
+        // If the latest block we have accepted by an authority is older than the current gc round,
+        // then do not attempt to fetch any blocks from that point as they will simply be skipped. Instead
+        // do attempt to fetch from the gc round.
+        if gc_enabled && last_received < gc_round {
+            warn!(
+                "Last received block for peer {peer} is older than GC round, {last_received} < {gc_round}, fetching from GC round"
```
[the remark this answers is not in the document]
Comment (akichidis) [reply consensus/core/src/subscriber.rs:72]: That was mostly me trying to be a little bit more verbose on the logging just to make sure that we will have historical logging  as debug/trace won't appear until on demand and it might be already late. I didn't consider this being too noisy , but I am happy to downgrade if you feel so.

File: consensus/core/src/subscriber.rs [hunk 6]
```diff
@@ -59,11 +59,20 @@ impl<C: NetworkClient, S: NetworkService> Subscriber<C, S> {
...
         let authority_service = self.authority_service.clone();
-        let last_received = self
-            .dag_state
-            .read()
-            .get_last_block_for_authority(peer)
-            .round();
+        let dag_state = self.dag_state.read();
+        let mut last_received = dag_state.get_last_block_for_authority(peer).round();
+        let gc_round = dag_state.gc_round();
+        let gc_enabled = dag_state.gc_enabled();
```
[the remark this answers is not in the document]
Comment (akichidis) [reply consensus/core/src/subscriber.rs:65]: You are right on the lock, but probably I'll go for something like this below as I would like to have the log line I've added

```suggestion
        let (mut last_received, gc_round, gc_enabled) = {
               let dag_state = self.dag_state.read();
               (dag_state.get_last_block_for_authority(peer).round(),  dag_state.gc_round(), dag_state.gc_enabled())
        }
```

File: consensus/core/src/block_manager.rs [hunk 7]
```diff
@@ -206,17 +241,47 @@ impl BlockManager {
...
+            return TryAcceptResult::Skipped;
+        }
+
+        // Keep only the ancestors that are greater than the GC round to check for their existence. Keep in mind that if GC is disabled
+        // then gc_round will be 0 and all ancestors will be considered.
+        let ancestors = if gc_enabled {
+            block
+                .ancestors()
+                .iter()
+                .filter(|ancestor| ancestor.round == GENESIS_ROUND || ancestor.round > gc_round)
```
[the remark this answers is not in the document]
Comment (mwtian) [reply consensus/core/src/block_manager.rs:275]: On second thought this logic is appropriate. If `gc_round` is defined to be the highest round not GC'ed we won't need to check for `ancestor.round == GENESIS_ROUND`. Also, I thought about checking `ancestor.round > gc_round || (ancestor.round == GENESIS_ROUND && gc_round == GENESIS_ROUND)` instead. Maybe it is an overkill but we may start to assume ancestors at or below gc_round are not checked in future.

File: consensus/core/src/subscriber.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mwtian) [reply consensus/core/src/subscriber.rs:76]: I think just info would be good.

File: consensus/core/src/block_manager.rs [hunk 8]
```diff
@@ -359,6 +419,66 @@ impl BlockManager {
...
+            assert!(!self.suspended_blocks.contains_key(block_ref), "Block should not be suspended, as we are causally GC'ing and no suspended block should exist for a missing ancestor.");
+
+            // Also remove it from the missing list - we don't want to keep looking for it.
+            self.missing_blocks.remove(block_ref);
+
+            // Find all the children blocks that have a dependency on this one and try to unsuspend them
+            let unsuspended_blocks = self.try_unsuspend_children_blocks(*block_ref);
+
+            unsuspended_blocks.iter().for_each(|block| {
+                if block.round() <= gc_round {
```
Comment (arun-koshy) [inline consensus/core/src/block_manager.rs:453]: Should this be filtering unsuspended blocks that have been gc'd?

File: consensus/core/src/core.rs [hunk 9]
```diff
@@ -731,6 +741,18 @@ impl Core {
...
+                            // When GC is enabled we want to include only ancestors that are:
+                            // 1) blocks of the genesis round
+                            // 2) blocks that are parents of the previous round (clock_round - 1)
+                            // 3) blocks that are parents earlier than the previous round, but higher than the gc_round
+                            return block.round() == GENESIS_ROUND
+                                || block.round() == clock_round - 1
```
Comment (arun-koshy) [inline consensus/core/src/core.rs:751]: why do we need `block.round() == clock_round - 1` isn't that covered by `block.round() > gc_round` or are you saying you will include a gc'd ancestor if it is from the parent round? That does not seem like a possible scenario

File: consensus/core/src/authority_node.rs [hunk 10]
```diff
@@ -487,12 +487,17 @@ mod tests {
...
+        #[values(0, 5, 10)] gc_depth: u32,
     ) {
         let db_registry = Registry::new();
         DBMetrics::init(&db_registry);
 
         const NUM_OF_AUTHORITIES: usize = 4;
         let (committee, keypairs) = local_committee_and_keys(0, [1; NUM_OF_AUTHORITIES].to_vec());
+        let mut protocol_config = ProtocolConfig::get_for_max_version_UNSAFE();
+        if gc_depth > 0 {
+            protocol_config.set_consensus_gc_depth_for_testing(gc_depth);
```
[the remark this answers is not in the document]
Comment (akichidis) [reply consensus/core/src/authority_node.rs:499]: Dah, absolute no reason - don't know what I did this. Thanks Arun!

File: consensus/core/src/block_manager.rs (hunk 8 above)
Comment (akichidis) [reply consensus/core/src/block_manager.rs:453]: hhm could you please elaborate a bit on this?

File: consensus/core/src/block_manager.rs [hunk 11]
```diff
@@ -413,6 +533,9 @@ enum TryAcceptResult {
     // The block has been processed before and already exists in BlockManager (and is suspended) or
     // in DagState (so has been already accepted). No further processing has been done at this point.
     Processed,
+    // When a received block is <= gc_round, then we simply skip its processing as there is no meaning
+    // do any action on it or even store it.
+    Skipped,
```
[the remark this answers is not in the document]
Comment (akichidis) [reply consensus/core/src/block_manager.rs:538]: Actually initially I had it as `Rejected` - but then I considered `Skipped` as the rejection probably conveys a more strong message (ex because there is something wrong with the block). Here we just choose to not process it as there is no meaning to.

File: consensus/core/src/core.rs (hunk 9 above)
Comment (akichidis) [reply consensus/core/src/core.rs:751]: Yeah I think that's me getting a bit paranoid with the checks here. We can certainly just do as you said `block.round() > gc_round`.  In fact the end check can just be:

```
if gc_enabled && gc_round > GENESIS_ROUND {
    return block.round() > gc_round;
}
```

File: consensus/core/src/block_manager.rs (hunk 8 above)
Comment (arun-koshy) [reply consensus/core/src/block_manager.rs:460]: How I am understanding this is that we have found a block in `missing_ancestors` that is garbage collected so we then `try_unsuspend_children_blocks` with this block no longer "missing". That will potentially return a set of blocks that have been unsuspended. You are then checking if any of those unsuspended blocks are <= gc_round which means they too should be garbage collected. 

So my confusion is why you only update the `total_blocks_gced` count but then accept these unsuspended blocks with `verify_block_timestamps_and_accept`?

Comment (akichidis) [reply consensus/core/src/block_manager.rs:460]: Oh got it. Ok maybe then the naming doesn't convey well the intention, but practically here I would like to capture "how many blocks under the gc_round have been accepted". Yes we do accept them in the end - which we could easily not to - but they can still be considered gc'ced. I can probably rename the variable to something like `blocks_accepted_below_gc_round`

File: consensus/core/src/block_manager.rs [hunk 12]
```diff
@@ -206,17 +241,47 @@ impl BlockManager {
...
+        let gc_enabled = dag_state.gc_enabled();
 
         // If block has been already received and suspended, or already processed and stored, or is a genesis block, then skip it.
         if self.suspended_blocks.contains_key(&block_ref) || dag_state.contains_block(&block_ref) {
             return TryAcceptResult::Processed;
         }
 
-        let ancestors = block.ancestors();
+        // If the block is <= gc_round, then we simply skip its processing as there is no meaning do any action on it or even store it.
+        if gc_enabled && block.round() <= gc_round {
```
[the remark this answers is not in the document]
Comment (akichidis) [reply consensus/core/src/block_manager.rs:260]: Let me revisit this on the upcoming PRs. Right now it's mostly this place where this is needed but if I can re-use I'll refactor.
